UnifiedPDF: Implement initial support for dynamic content scaling

kkinnunen-apple opened the pull request:
#### 59ff4f6298f32fe862e26734d2707b8a0f4beae4
<pre>
UnifiedPDF: Implement initial support for dynamic content scaling
<a href="https://bugs.webkit.org/show_bug.cgi?id=287623">https://bugs.webkit.org/show_bug.cgi?id=287623</a>
<a href="https://rdar.apple.com/144775149">rdar://144775149</a>

Reviewed by Abrar Rahman Protyasha, Tim Horton, and Simon Fraser.

Add PlatformCALayerClient::platformCALayerDynamicContentScalingDisplayList(),
a function to override when the layer contents provides the dynamic
content scaling display lists in explicit fashion.

Implement this for tiled layers that provide the PDF content through
AsyncPDFRenderer. Map the platform layer to a particular tile in a tile
grid, and map that tile into rendered PDF content. Make the PDF content,
RenderedPDFTile, to hold the DCS display list.

Generate the DCS display list by calling the PDF drawing code twice:
once with a bitmap-backed ImageBuffer, once with DCS display list
-backed ImageBuffer. The bifurcated ImageBuffer would not work, as
the PDF drawing uses the underlying platform context that is not
bifurcated.

* Source/WebCore/platform/graphics/TiledBacking.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayerClient.h:
(WebCore::PlatformCALayerClient::platformCALayerDynamicContentScalingDisplayList const):
* Source/WebCore/platform/graphics/ca/TileController.cpp:
(WebCore::TileController::dynamicContentScalingDisplayListForTile):
* Source/WebCore/platform/graphics/ca/TileController.h:
* Source/WebCore/platform/graphics/ca/TileGrid.cpp:
(WebCore::TileGrid::platformCALayerDynamicContentScalingDisplayList const):
* Source/WebCore/platform/graphics/ca/TileGrid.h:
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingImageBufferBackend.h:
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingImageBufferBackend.mm:
(WebKit::DynamicContentScalingImageBufferBackend::createBackendHandle const):
(WebKit::DynamicContentScalingImageBufferBackend::displayList const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
(WebKit::RemoteLayerBackingStore::displayListHandle const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStoreProperties::applyBackingStoreToLayer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
The display list was incorrectly serialized as ImageBufferBackendHandle
a std::variant containing also DynamicContentScalingDisplayList. Instead,
mark it directly as WebCore::DynamicContentScalingDisplayList.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm:
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::displayListHandle const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::displayListHandle const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.mm:
(WebKit::AsyncPDFRenderer::releaseMemory):
(WebKit::AsyncPDFRenderer::dynamicContentScalingDisplayListForTile):
(WebKit::renderPDFTile):
(WebKit::renderPDFTileToImage):
(WebKit::renderPDFTileToDynamicContentScalingDisplayList):
(WebKit::AsyncPDFRenderer::serviceRequestQueues):
(WebKit::AsyncPDFRenderer::didCompleteTileRender):
(WebKit::AsyncPDFRenderer::dynamicContentScalingResourceCache):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.mm:
(WebKit::PDFDiscretePresentationController::layerAllowsDynamicContentScaling const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFScrollingPresentationController.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFScrollingPresentationController.mm:
(WebKit::PDFScrollingPresentationController::layerAllowsDynamicContentScaling const):

Canonical link: <a href="https://commits.webkit.org/291019@main">https://commits.webkit.org/291019@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/35f02a704788f9b1fcdbe6d8237170a4c026de01

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91553 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11084 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/666 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96522 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42239 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93603 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11462 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19767 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70312 "Failure limit exceed. At least found 7 new test failures: imported/w3c/web-platform-tests/css/css-animations/CSSAnimation-pausing.tentative.html imported/w3c/web-platform-tests/css/css-animations/CSSAnimation-ready.tentative.html imported/w3c/web-platform-tests/css/css-animations/CSSAnimation-startTime.tentative.html imported/w3c/web-platform-tests/css/css-animations/Element-getAnimations-dynamic-changes.tentative.html imported/w3c/web-platform-tests/css/css-animations/Element-getAnimations.tentative.html imported/w3c/web-platform-tests/css/css-animations/animate-with-color-mix.html imported/w3c/web-platform-tests/css/css-animations/animation-before-initial-box-construction-001.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27827 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94554 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8755 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82992 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50636 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8518 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/581 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41409 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78838 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/584 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98526 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18714 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13953 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79338 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18969 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78830 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78542 "Found 1 new API test failure: /WebKitGTK/TestInspector:/webkit/WebKitWebInspector/manual-attach-detach (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19468 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23060 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/447 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/11843 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18710 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23986 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18421 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21881 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20187 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/platform/graphics/TiledBacking.h:
```diff
@@ -33,6 +33,10 @@
 #include <wtf/WeakPtr.h>
 
+#if ENABLE(RE_DYNAMIC_CONTENT_SCALING)
+#include "DynamicContentScalingDisplayList.h"
```
Nit: you can forward declare `class DynamicContentScalingDisplayList;` since `std::optional` doesn't need the full type definition for the return type.

aprotyas replied:
Same comment about some of the other header files

kkinnunen-apple replied:
No, I don't think that's correct.
std::optional doesn't play a role here.

These behave the same way:
```
std::optional<DynamicContentScalingDisplayList> a();
DynamicContentScalingDisplayList b();
```

If your rule was correct rule, mostly nothing would need to be included.

The fact is that the compiler needs the type to be complete when the function is called, in both cases.

The convention is that it's poor form to expect the caller to understand that they would need to include "DynamicContentScalingDisplayList" before calling.

Thus the header declaring a function returning DynamicContentScalingDisplayList by value should include the header. Returning it through std::optional is still returning it by value.

smfr commented on Source/WebCore/platform/graphics/ca/TileGrid.cpp:
```diff
@@ -814,6 +814,17 @@ bool TileGrid::platformCALayerNeedsPlatformContext(const PlatformCALayer* layer)
+#if ENABLE(RE_DYNAMIC_CONTENT_SCALING)
+std::optional<DynamicContentScalingDisplayList> TileGrid::platformCALayerDynamicContentScalingDisplayList(const PlatformCALayer* layer) const
+{
+    for (auto& [tileIndex, tileInfo] : m_tiles) {
```
It's not great that we have to linear search through the tiles.

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFScrollingPresentationController.mm:
```diff
@@ -392,6 +392,14 @@
 }
 
+#if ENABLE(RE_DYNAMIC_CONTENT_SCALING)
+bool PDFScrollingPresentationController::layerAllowsDynamicContentScaling(const GraphicsLayer*) const
```
I fully expect that making "allows" a lie is going to come back and bite me someday. Maybe we need to rename this to be more explicitly about recording DCS display lists during painting? I'm not sure what exactly to go with.

kkinnunen-apple committed:
UnifiedPDF: Implement initial support for dynamic content scaling
https://bugs.webkit.org/show_bug.cgi?id=287623
rdar://144775149

Reviewed by Abrar Rahman Protyasha, Tim Horton, and Simon Fraser.

Add PlatformCALayerClient::platformCALayerDynamicContentScalingDisplayList(),
a function to override when the layer contents provides the dynamic
content scaling display lists in explicit fashion.

Implement this for tiled layers that provide the PDF content through
AsyncPDFRenderer. Map the platform layer to a particular tile in a tile
grid, and map that tile into rendered PDF content. Make the PDF content,
RenderedPDFTile, to hold the DCS display list.

Generate the DCS display list by calling the PDF drawing code twice:
once with a bitmap-backed ImageBuffer, once with DCS display list
-backed ImageBuffer. The bifurcated ImageBuffer would not work, as
the PDF drawing uses the underlying platform context that is not
bifurcated.

* Source/WebCore/platform/graphics/TiledBacking.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayerClient.h:
(WebCore::PlatformCALayerClient::platformCALayerDynamicContentScalingDisplayList const):
* Source/WebCore/platform/graphics/ca/TileController.cpp:
(WebCore::TileController::dynamicContentScalingDisplayListForTile):
* Source/WebCore/platform/graphics/ca/TileController.h:
* Source/WebCore/platform/graphics/ca/TileGrid.cpp:
(WebCore::TileGrid::platformCALayerDynamicContentScalingDisplayList const):
* Source/WebCore/platform/graphics/ca/TileGrid.h:
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingImageBufferBackend.h:
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingImageBufferBackend.mm:
(WebKit::DynamicContentScalingImageBufferBackend::createBackendHandle const):
(WebKit::DynamicContentScalingImageBufferBackend::displayList const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
(WebKit::RemoteLayerBackingStore::displayListHandle const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStoreProperties::applyBackingStoreToLayer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
The display list was incorrectly serialized as ImageBufferBackendHandle
a std::variant containing also DynamicContentScalingDisplayList. Instead,
mark it directly as WebCore::DynamicContentScalingDisplayList.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm:
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::displayListHandle const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::displayListHandle const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.mm:
(WebKit::AsyncPDFRenderer::releaseMemory):
(WebKit::AsyncPDFRenderer::dynamicContentScalingDisplayListForTile):
(WebKit::renderPDFTile):
(WebKit::renderPDFTileToImage):
(WebKit::renderPDFTileToDynamicContentScalingDisplayList):
(WebKit::AsyncPDFRenderer::serviceRequestQueues):
(WebKit::AsyncPDFRenderer::didCompleteTileRender):
(WebKit::AsyncPDFRenderer::dynamicContentScalingResourceCache):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.mm:
(WebKit::PDFDiscretePresentationController::layerAllowsDynamicContentScaling const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFScrollingPresentationController.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFScrollingPresentationController.mm:
(WebKit::PDFScrollingPresentationController::layerAllowsDynamicContentScaling const):

Canonical link: https://commits.webkit.org/291019@main
